Fixing a bug in .to for qtensors so scale/zp move too

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61576**

Summary: This also fixed an issue in the
empty_quantized_per_channel_affine function where specifying a device
that was different from the device of scale/zp would result in a
mismatched qtensor with the data being on one device and the scale/zp tensors would be on another (related bug)

Test Plan: python test/test_quantization.py
testquantizedtensor.test_per_channel_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29675461](https://our.internmc.facebook.com/intern/diff/D29675461)

HDCharles committed:
Fixing a bug in .to for qtensors so scale/zp move too

Summary: This also fixed an issue in the
empty_quantized_per_channel_affine function where specifying a device
that was different from the device of scale/zp would result in a
mismatched qtensor

Test Plan: python test/test_quantization.py
testquantizedtensor.test_per_channel_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29675461).

HDCharles committed:
Update on "Fixing a bug in .to for qtensors so scale/zp move too"


Summary: This also fixed an issue in the
empty_quantized_per_channel_affine function where specifying a device
that was different from the device of scale/zp would result in a
mismatched qtensor with the data being on one device and the scale/zp tensors would be on another (related bug)

Test Plan: python test/test_quantization.py
testquantizedtensor.test_per_channel_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29675461](https://our.internmc.facebook.com/intern/diff/D29675461)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29675461).

jerryzh168 commented on aten/src/ATen/native/quantized/TensorFactories.cpp:
```diff
@@ -58,8 +58,20 @@ Tensor empty_per_channel_affine_quantized(
+    new_scales=scales.cpu();
+    new_zero_points=zero_points.cpu();
+  } else {
+    new_scales=scales.cuda();
```
can we get the device from options.device()?
e.g. does
```
new_scales = scales.to(options.device());
```
work?

HDCharles replied:
no, at least I couldn't get that to work, all the .to methods require more arguments. The only thing I could get working was the .cpu() and .cuda() functions. I even tried copying the syntax directly from 

https://github.com/pytorch/pytorch/blob/a07b08136fb982a85b24ef09b86a89496247b1a3/aten/src/ATen/templates/TensorBody.h#L643

not sure what the issue is/was.

HDCharles replied:
nvm, I'm not sure what I tested before. This has been changed

HDCharles committed:
Update on "Fixing a bug in .to for qtensors so scale/zp move too"


Summary: This also fixed an issue in the
empty_quantized_per_channel_affine function where specifying a device
that was different from the device of scale/zp would result in a
mismatched qtensor with the data being on one device and the scale/zp tensors would be on another (related bug)

Test Plan: python test/test_quantization.py
testquantizedtensor.test_per_channel_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29675461](https://our.internmc.facebook.com/intern/diff/D29675461)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29675461).

HDCharles committed:
Update on "Fixing a bug in .to for qtensors so scale/zp move too"


Summary: This also fixed an issue in the
empty_quantized_per_channel_affine function where specifying a device
that was different from the device of scale/zp would result in a
mismatched qtensor with the data being on one device and the scale/zp tensors would be on another (related bug)

Test Plan: python test/test_quantization.py
testquantizedtensor.test_per_channel_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29675461](https://our.internmc.facebook.com/intern/diff/D29675461)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29675461).

jerryzh168 commented on aten/src/ATen/native/quantized/TensorFactories.cpp:
```diff
@@ -59,7 +59,7 @@ Tensor empty_per_channel_affine_quantized(
       "Must provide data type for Tensor creation functions.");
   QuantizerPtr quantizer = make_per_channel_affine_quantizer(
-          scales, zero_points, axis, typeMetaToScalarType(options.dtype()));
+          scales.to(options.device()), zero_points.to(options.device()), axis, typeMetaToScalarType(options.dtype()));
```
btw, is this the only place we need this?

HDCharles replied:
its only needed in per_channel and everything I'm seeing flows through this specific constructor

HDCharles committed:
Update on "Fixing a bug in .to for qtensors so scale/zp move too"


Summary: This also fixed an issue in the
empty_quantized_per_channel_affine function where specifying a device
that was different from the device of scale/zp would result in a
mismatched qtensor with the data being on one device and the scale/zp tensors would be on another (related bug)

Test Plan: python test/test_quantization.py
testquantizedtensor.test_per_channel_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29675461](https://our.internmc.facebook.com/intern/diff/D29675461)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29675461).

HDCharles committed:
Update on "Fixing a bug in .to for qtensors so scale/zp move too"


Summary: This also fixed an issue in the
empty_quantized_per_channel_affine function where specifying a device
that was different from the device of scale/zp would result in a
mismatched qtensor with the data being on one device and the scale/zp tensors would be on another (related bug)

Test Plan: python test/test_quantization.py
testquantizedtensor.test_per_channel_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29675461](https://our.internmc.facebook.com/intern/diff/D29675461)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29675461).
